module/zfs/bpobj.c

```diff
@@ -242,19 +280,23 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
 		ASSERT3U(offset, >=, dbuf->db_offset);
 		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
 
-		bparray = dbuf->db_data;
-		bp = &bparray[blkoff];
+		blkptr_t *bparray = dbuf->db_data;
+		blkptr_t *bp = &bparray[blkoff];
 		err = func(arg, bp, tx);
 		if (err)
 			break;
+
 		if (free) {
-			bpo->bpo_phys->bpo_bytes -=
-			    bp_get_dsize_sync(dmu_objset_spa(bpo->bpo_os), bp);
+			spa_t *spa = dmu_objset_spa(bpo->bpo_os);
+			comp_freed = uncomp_freed = 0;
+			freed = bp_get_dsize_sync(spa, bp);
```

Comment:
freed, comp_freed, uncomp_freed should be declared here (but see below about calling propagate_space_reduction once, in which case we would want to use them to accumulate the sum for all the bp's we iterate over, in which case they would need to continue to be declared (and initialized!) in the outer scope).